src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -348,6 +364,48 @@ private void EventLoop()
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void ScheduleToProcessEvents()
+        {
+            // Schedule a thread pool work item to process events. Only one work item is scheduled at any given time to avoid
+            // over-parallelization. When the work item begins running, this field is reset to 0, allowing for another work item
+            // to be scheduled for parallelizing processing of events.
+            if (Interlocked.CompareExchange(ref _eventQueueProcessingRequested, 1, 0) == 0)
+            {
+                ThreadPool.UnsafeQueueUserWorkItem(this, preferLocal: false);
+            }
+        }
+
+        void IThreadPoolWorkItem.Execute()
+        {
+            // Indicate that a work item is no longer scheduled to process events. The change needs to be visible to enqueuer
+            // threads (only for EventLoop() currently) before an event is attempted to be dequeued. In particular, if an
+            // enqueuer queues an event and does not schedule a work item because it is already scheduled, and this thread is
+            // the last thread processing events, it must see the event queued by the enqueuer.
+            Interlocked.Exchange(ref _eventQueueProcessingRequested, 0);
+
+            ConcurrentDictionary<IntPtr, SocketAsyncContext> handleToContextMap = _handleToContextMap;
+            ConcurrentQueue<Event> eventQueue = _eventQueue;
+            while (eventQueue.TryDequeue(out Event ev))
+            {
+                handleToContextMap.TryGetValue(ev.Handle, out SocketAsyncContext? context);
+                if (context == null)
+                {
+                    continue;
+                }
+
+                // An event was successfully dequeued, and as there may be more events to process, speculatively schedule a work
+                // item to parallelize processing of events. Since this is only for additional parallelization, doing so
+                // speculatively is ok.
```

Comment:
Maybe worth mentioning explicitly that the parallelization makes it impossible for continuations to block one another.

Reply:
I don't think it's possible for a continuation to block on another continuation for the same socket even without the parallelization done here. If a continuation blocks on a synchronous socket operation the next epoll event that serves that blocking operation would schedule another work item. The parallelization done here is only for making use of more procs.

Reply:
> I don't think it's possible for a continuation to block on another continuation for the same socket even without the parallelization done here.

It may be on different sockets. And we shouldn't reply on the next epoll_wait with events to get things moving.

Reply:
If parallelization is not done here, and if a blocking socket operation is performed on this thread as a result of running a user continuation (on any socket), then the set of IO events that are already queued would not be releasing that blocking operation anyway since it's a new operation that got added. Only another epoll event would get that blocking operation moving (before this change too).

However, if it's some other kind of blocking related to other already-queued, for example if a user continuation blocks waiting for another already-in-progress socket operation to complete, then the parallelization done here would help with unblocking that more quickly, though not guaranteed and can still lead to thread pool starvation issues, and there would have already been potential thread pool starvation issues before this change from those kinds of blocking.

I'm trying to understand if there would be a correctness issue from not parallelizing here. If there is a possibility of correctness issue then it may be necessary to queue up a replica before the loop instead.

Reply:
To alleviate potential issues from other kinds of blocking in user callbacks maybe it would be safer to ensure there is a queued replica (non-speculatively) if it has not yet scheduled a replica.

Reply:
Made the change I mentioned above in the latest update